### ISSUE TITLE
git_patch_to_buf Does not work if underlying blob is freed

### ISSUE THREAD

**Issue by brandonio21** [libgit2/libgit2#4442]
This issue was discovered in pygit2 https://github.com/libgit2/pygit2/issues/754 and there are some pretty detailed reproduction steps there.

# Reproductions steps
Create a patch object from at least one blob. Free the blob. Print the patch to a buffer. The patch will be mangled because the underlying blob was freed.

# Expected behavior
A patch created from a blob should be able to print to a buffer without issue when the blob is freed.

In short, the patch should not need the blob to exist for it to print. The patch should hold a reference to the blob.

**Comment by ethomson** [comment]
I agree - this is unfortunate as it's neither obvious nor documented.  But indeed this is the case at the moment.

We should either clearly document this or find a solution where we refcount the blob or blob data.  I'm in favor of the latter.

**Comment by pks-t** [comment]
All our objects are already refcounted. We simply fail to use the mechanism correctly here. A PR will follow shortly

**Comment by brandonio21** [comment]
Fixed in #4447 :)
